### DISCUSSION TITLE
Os conhecimentos em Engenharia de Software fazem sentido no mundo Front End?

### DISCUSSION THREAD

**Discussion by woliveiras** [frontendbr/forum#568] [Discussão]
Galera, ontem escrevi um post com os tópicos que separei para estudar a respeito de Engenharia de Software e um amigo comentou que não precisava estudar tudo aquilo.

O post foi esse aqui: [Os caminhos da Engenharia de Software: o que estudar](https://woliveiras.com.br/posts/os-caminhos-da-engenharia-de-software-o-que-estudar)

Eu, pessoalmente, quero estudar esses tópicos por que eu gosto desse assunto e sei que muito disso me ajuda a ser bom programador JavaScript, mas me levantou a seguinte dúvida: 

**Como desenvolvedores front end, quais os tópicos de matérias de Ciência da Computação e Engenharia precisamos saber?**

**Comment by wmartins** [comment]
A questão do que estudar vai depender mais do problema em que se quer resolver. Para uma pessoa que tenha que desenvolver um monte de telas, reutilizáveis, extensíveis e etc., talvez faça mais sentido saber muito bem conceitos de design patterns ao invés de ter uma boa base em matemática, estruturas de dados e afins (embora ambos sejam vistos em boa parte dos currículos de Ciência/Engenharia da Computação).

Tem um vídeo que eu acho fantástico, de uma JSConf em 2014, onde o palestrante fala um pouco sobre essa relação de front end, ciência da computação, algoritmos e etc:

https://www.youtube.com/watch?v=90NsjKvz9Ns

Vale dar uma olhada!

**Comment by felipeuntill** [comment]
# TLDR;

Aprendi a programar ainda enquanto criança e em uma época onde o front-end se resumia a layout slicing e tables, ou seja nada muito interessante. Porém com o passar dos anos me deparei com o tableless e javascript, tecnologias que me proporcionaram amor imediato pelo mundo front-end parei de depreender meus esforços como "programador" e decidi que minha vida seria fazer aquilo.

Ai que começa o problema, por ter atuado por muitos anos com desenvolvimento backend principalmente com Java e C#(Bancos) eu aprendi a abstrair tudo em design patterns que no backend são impreteríveis porém no front-end são over engineering.

Sempre me pego tentando adaptar abstrações oriundas de outras vertentes computacionais para o desenvolvimento front-end e entrando em um ciclo de complexidade desnecessário ex:

1-) Hoje em dia considerando que todo programador backend experiente utiliza os fundamentos de SOLID em seus projetos, ignorar a inversão de controle é algo praticamente inaceitável no meu subconciente, então sempre procuro meios de fazer IOC no js inclusive criei uma biblioteca para tal chamada [inverter-ioc](https://github.com/felipeuntill/inverter-ioc), agora ai que entra o problema consultando projetos open source desenvolvido pela nata do frontend, ninguém faz inversão de controle então porque eu deveria trazer este tipo de abstração para o front-end.

2-) Design Patterns (GOF) sempre que possível tento implementar design patterns no meu código porém novamente apesar de possuirem valor agregado no nodejs, não vejo pessoas procurando soluções similares no front-end.

Dados os argumentos acima, minha dúvida é, será que por ter sido backend primeiro eu não consigo ter cabeça aberta? ou conhecer os paradigmas de engenharia de projeto faz com que as pessoas fiquem engessadas?

**Comment by fititnt** [comment]
**Interação Humano-Computador** é a disciplina mais importante para diferenciar dev front-end de back-end.

A própria diferenciação entre desenvolvedor front-end e back-end deveria (até onde era ideia anos atras) **ser para dar mais espaço para o desenvolvedor se focar na interface**, afinal, desenvolvedor front-end nasceu da divisão que era antes chamado de, apenas, desenvolvedor/programador web.

Se fosse escolher apenas três disciplinas de ECP e CiC para sugerir, seriam, a ponto de até mesmo uma pessoa não matriculada ter cara de pau de ir assistir aulas, mesmo sem fazer vestibular, seriam nessa ordem de indicação: 1) **alguma de introdução a programação** (voltada para cursos de ECP ou CiC, não as que são para outros cursos); 2) **Interação Humano-Computador**; 3)  alguma de outro curso da universidade, provavelmente artes ou design de produto, e tenha uma **pegada focada ser um semestre apenas sobre temas como teoria de cores, formas geométricas e tipografia na forma como é percebida pelo humano**.

As duas últimas ali geralmente seriam mal valorizadas DENTRO de uma universidade no círculo de colegas, mais por bully de colegas e de professores na ideia de "quanto código mais complicado e baix nível, melhor", do que por falta de aplicabilidade no mundo real, em um contexto de desenvolvedor web que faça interfaces.

**Comment by felquis** [comment]
Sim!

**Comment by felipefialho** [comment]
Faz, e faz pra carario.

**Comment by kivervinicius** [comment]
Lembrei daquelas cartas de ressurreição de tópicos que colocavam nos fórums de antigamente kkk

**Comment by woliveiras** [comment]
![REVIVEU](http://1.bp.blogspot.com/-oOYd8saLnJQ/UBczYGsSFeI/AAAAAAAAAC8/1ir50lPzkvs/s1600/MonsterRebornDPYG-EN-R-1E.jpg)
